Re-gigger backfilling `technology_description` & make `prime_mover_code` an annually harvested column

cmgosnell opened the pull request:
Update: I incorporated a change of `prime_mover_code` from a "static" -> "annual" harvested/normalized column!!

The previous fill-in methodology was filling in the `technology_description` by backfilling and then using the unique mapping between ESC and tech type. This PR edits the unique mapping fill-in portion to ALSO include the `prime_mover_code`.

Because this suggestion includes a two-column:one-column map, I had to convert the `map(dict)` methodology into a split-apply via merge-combine methodology.

## how I got here...
I came across this by looking at the older years of `plant_id_eia == 1961`. this generator was converted into a gas plant so the backfilling didn't work. And this was an `energy_source_code_1 == "BIT"` generator. `BIT` has either mapped to `Conventional Steam Coal` or `Coal Integrated Gasification Combined Cycle` tech types. This is not a unique mapping so no `BIT`'s were getting filled in. 

Out of 135 potential esc/pm -> tech mappings, only 7 of them had non-unique map:
```
test = (
    gens.groupby(["energy_source_code_1", "prime_mover_code"])
    [["technology_description"]]
    .nunique()
)
len(test)
> 135
```

![image](https://user-images.githubusercontent.com/25487998/166016623-f5b6e22a-b2b9-4812-b2b8-7817b00db3a7.png)

aesharpe commented on src/pudl/output/eia860.py:
```diff
@@ -368,13 +367,15 @@ def generators_eia860(
 
 def fill_generator_technology_description(gens_df: pd.DataFrame) -> pd.DataFrame:
-    """Fill in missing ``technology_description`` based on generator and energy source.
+    """Fill in missing ``technology_description`` based by backfilling & unquie mapping.
```
Tiny comment, but unique is spelled wrong

cmgosnell replied:
lol of course ty!

aesharpe commented on src/pudl/output/eia860.py:
```diff
@@ -368,13 +367,15 @@ def generators_eia860(
+    unique mappings that are observed between ``energy_source_code_1``,
+    ``prime_mover_code`` and ``technology_type`` across all years and generators.
 
     As a result, more than 95% of all generator records end up having a
     ``technology_description`` associated with them.
```
Do you know what the coverage is now that you've integrated `prime_mover_code`?

cmgosnell replied:
ha its 97%... which is a HUGE improvement from 96%.

cmgosnell replied:
with moving the PM code....  this is now 98.1% 😎

aesharpe reviewed (commented):
This looks robust to me! Just a few bb comments

cmgosnell commented:
I added another layer of this PR. Now, ontop of the backfilling of the `technology_description`, I've also moved the `prime_mover_code` from a "static" to an "annual" harvested column. See details on why from [this comment](https://github.com/catalyst-cooperative/pudl/issues/1585#issuecomment-1116150377)

cmgosnell commented:
ALSO i still need to run `tox -e nuke` to run all the validation tests before merging.... this feels important given the change in the table structure. (even though - believe it or not - there seem to be NO changes required in the output layer)

cmgosnell commented:
Okay @aesharpe I changed a few things and ran all of the validation tests and they check out!!

aesharpe commented:
> Update: I incorporated a change of `prime_mover_code` from a "static" -> "annual" harvested/normalized column!!

Does this mean that before now we were overwriting certain PM values based on whatever we decided the "static" value was?

aesharpe commented:
> this feels important given the change in the table structure.

Which table structure changes? The annual table? The output tables would only change insofar as the PM column differs based on the new annual data, right, so no structure changes there?

cmgosnell commented:
@aesharpe good questions

> Does this mean that before now we were overwriting certain PM values based on whatever we decided the "static" value was?

Yes. 98% of PM's never change ever so we lumped this columns into the "static" (i.e. you never change ever and any change is probably a data entry issue)... but withing that 2% there were some clear instances of *actual* change.

> Are there any tables that might get messed up by the addition of PM to annual? Or new documentation to add?

Fortunately, the only mode of access to these tables that we've baked in is through `pudl.output.eia860.generators_eia860`. And this output function was always merging all of the columns of the annually harvested tables and the static harvested generator tables together. So weirdly no output methods needed updating. I was vv surprised but that is the niceness of layering our access via the output tables.

On documentation, the docs should be built off of the metadata and that is now updated. This change should *definitely* be added to release notes when we make a new release. I've already flagged this for @zaneselvans... (should we make a running `0.7.0 (2022-XX-XX)` section??)

> Which table structure changes? The annual table? 

A column (`prime_mover_code`) got moved from the static table (`generators_entity_eia`) to the annual table (`generators_eia860`). 

> The output tables would only change insofar as the PM column differs based on the new annual data, right, so no structure changes there?

Exactly. The structure  of `pudl_out.gens_eia80()` is exactly the same. Some of the PM codes of those 2% of gens that differ in time will be different.

zaneselvans commented:
Yes, if there isn't already a v0.7.0 section in the release notes definitely add one. We should make edits to the release notes in every significant PR.
